wdl-lint/src/rules/disallowed_declaration_name.rs

```diff
@@ -0,0 +1,199 @@
+//! A lint rule that disallows declaration names with type suffixes.
+
+use std::collections::HashSet;
+
+use wdl_ast::AstNode;
+use wdl_ast::AstToken;
+use wdl_ast::Diagnostic;
+use wdl_ast::Diagnostics;
+use wdl_ast::Document;
+use wdl_ast::Span;
+use wdl_ast::SupportedVersion;
+use wdl_ast::SyntaxElement;
+use wdl_ast::SyntaxKind;
+use wdl_ast::VisitReason;
+use wdl_ast::Visitor;
+use wdl_ast::v1::BoundDecl;
+use wdl_ast::v1::Decl;
+use wdl_ast::v1::PrimitiveTypeKind;
+use wdl_ast::v1::Type;
+use wdl_ast::v1::UnboundDecl;
+
+use crate::Rule;
+use crate::Tag;
+use crate::TagSet;
+
+/// A rule that identifies declaration names that include their type names as a
+/// suffix.
+#[derive(Debug, Default)]
+pub struct DisallowedDeclarationNameRule;
+
+/// Create a diagnostic for a declaration identifier that contains its type name
+fn decl_identifier_with_type(span: Span, decl_name: &str, type_name: &str) -> Diagnostic {
+    Diagnostic::warning(format!(
+        "declaration identifier '{decl_name}' ends with type name '{type_name}'",
+    ))
+    .with_rule("DisallowedDeclarationName")
+    .with_highlight(span)
+    .with_fix("rename the identifier to not include the type name")
+}
+
+impl Rule for DisallowedDeclarationNameRule {
+    fn id(&self) -> &'static str {
+        "DisallowedDeclarationName"
+    }
+
+    fn description(&self) -> &'static str {
+        "Disallows declaration names that end with their type name"
+    }
+
+    fn explanation(&self) -> &'static str {
```

Comment:
Can you update this text to include that an exception is made for `String`, `File`, and user-defined structs?